### PR TITLE
Article/seeds of inception 2

### PR DESCRIPTION
**Pull request by mechanicalpete**

### CODE REVIEW

File: _posts/2018-04-01-inception-pipelines-pt2.markdown [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+image: img/inception-pipelines/seed_germination.png
+---
+
+**Part 2 sprouting some website goodness** - A practical example for DevOps on AWS
+
+![sprouting some website goodness]({{ site.url }}/img/inception-pipelines/seed_germination_2.png)
+
+## What's the Problem
+
+If you've read my [first post](https://mechanicalrock.github.io//aws/continuous/deployment/2018/03/01/inception-pipelines-pt1) (and you have, haven't you), you either thought "that's absolute crap, why would I bother" or "hey, that's pretty neat, but what can I do with it". If you were the former, then avert your eyes because this post is targeted firmly at the latter.
```
Comment (jgunnink) [inline _posts/2018-04-01-inception-pipelines-pt2.markdown]: Each of the "thoughts" you have needs a `?`

File: _posts/2018-04-01-inception-pipelines-pt2.markdown [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+1. The first one to create is an [AWS Certificate Manager](https://aws.amazon.com/certificate-manager/) managed SSL certificate in the `us-east-1` region. This is the only region that CloudFront will look for the certificate. So, unless you're creating the stack in `us-east-1`, you're going to need to create it yourself. This also gives you the option to verify the certificate by email or DNS (email is the only option when it is created via CloudFormation). You can find out more [here](https://docs.aws.amazon.com/acm/latest/userguide/gs-acm-request.html).
+1. The second manual step is to ensure you have a Route 53 Hosted Zone. All that you need is the `HostedZoneId` as it is a parameter to the CloudFormation template discussed below.
+
+An optional step, at least until September 2017, is to create a [DNS CAA record](https://blog.qualys.com/ssllabs/2017/03/13/caa-mandated-by-cabrowser-forum). Having the `CAA` record helps if you are attempting to get a good grade on services like [SSL Labs](https://www.ssllabs.com/).
+
+Incidentally, using the CloudFormation template discussed in this post will get you an 'A' grade.
+
+## How It All Works
+
+What this template does for you is:
```
Comment (jgunnink) [inline _posts/2018-04-01-inception-pipelines-pt2.markdown]: `is` is redundant.

File: _posts/2018-04-01-inception-pipelines-pt2.markdown [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+### What Are the Files
+
+|File|Description|
+|----|-----------|
+|aws_infrastructure.yml|The magnum opus; the CloudFormation template that makes it all work|
+|aws_infrastructure.json|These are the parameters used by the CloudFormation template during execution|
+|aws_seed.yml|Gets a new CloudFormation deployment action snippet as described below which executes the `aws_infrastructure.yml` template.|
+
+## Taking It for A Spin
```
Comment (jgunnink) [inline _posts/2018-04-01-inception-pipelines-pt2.markdown]: inconsistent capitalisation

File: _posts/2018-04-01-inception-pipelines-pt2.markdown [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+  * Redirects `403` and `404` errors to the `index.html` page. This is needed for single page applications (like AngularJS/Angular) that create browser friendly URLs that do not map to S3 files. The high-level flow is the S3 gives `403`, CloudFront serves the `index.html` which kicks off the SPA that reads the URI and starts serving the right content.
+  * Send access logs to an S3 bucket.
+  * Uses an existing SSL certificate, and makes sure everything is accessing it via HTTPS
+* Creates a DNS 'A' record in the specified Route 53 hosted zone. This hides the CloudFront domain behind your friendly domain name.
+
+## Where Do I Get the Seed Files
+
+The files are on the Part 2 branch in the [GitHub repository](https://github.com/MechanicalRock/InceptionPipeline/tree/post/part-2).
+
+### What Are the Files
```
Comment (jgunnink) [inline _posts/2018-04-01-inception-pipelines-pt2.markdown]: inconsistent capitalisation

File: _posts/2018-04-01-inception-pipelines-pt2.markdown [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+* Creates an S3 bucket for holding the website files. These files do not need to be publicly accessible. In fact, we hide them away in the S3 Bucket Policy so that only CloudFront can serve them.
+* Creates an S3 bucket to hold the CloudFront access logs.
+* Creates a CloudFront distribution that sits in front of the S3 website bucket.
+  * Creates a CloudFront origin access identity so that no-one can access the S3 bucket outside of CloudFront
+  * Redirects `403` and `404` errors to the `index.html` page. This is needed for single page applications (like AngularJS/Angular) that create browser friendly URLs that do not map to S3 files. The high-level flow is the S3 gives `403`, CloudFront serves the `index.html` which kicks off the SPA that reads the URI and starts serving the right content.
+  * Send access logs to an S3 bucket.
+  * Uses an existing SSL certificate, and makes sure everything is accessing it via HTTPS
+* Creates a DNS 'A' record in the specified Route 53 hosted zone. This hides the CloudFront domain behind your friendly domain name.
+
+## Where Do I Get the Seed Files
```
Comment (jgunnink) [inline _posts/2018-04-01-inception-pipelines-pt2.markdown]: inconsistent capitalisation

File: _posts/2018-04-01-inception-pipelines-pt2.markdown [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+In this post I will be covering how to extend an [Inception Pipeline](https://github.com/MechanicalRock/InceptionPipeline/tree/master) to do something useful. In this instance, it is creating the infrastructure to host a single page application. On the projects I'm currently involved with, this is always the first piece of infrastructure we need (well, after first inceptioning up the pipeline).
+
+## What Technologies Are We Going to Use
+
+* [AWS Certificate Manager](https://aws.amazon.com/certificate-manager/)
+* [CloudFront](https://aws.amazon.com/cloudfront/)
+* [Route 53](https://aws.amazon.com/route53/)
+* [S3](https://aws.amazon.com/s3/)
+
+## What Are the Prerequisites
```
Comment (jgunnink) [inline _posts/2018-04-01-inception-pipelines-pt2.markdown]: inconsistent capitalisation

File: _posts/2018-04-01-inception-pipelines-pt2.markdown [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+
+![sprouting some website goodness]({{ site.url }}/img/inception-pipelines/seed_germination_2.png)
+
+## What's the Problem
+
+If you've read my [first post](https://mechanicalrock.github.io//aws/continuous/deployment/2018/03/01/inception-pipelines-pt1) (and you have, haven't you), you either thought "that's absolute crap, why would I bother" or "hey, that's pretty neat, but what can I do with it". If you were the former, then avert your eyes because this post is targeted firmly at the latter.
+
+In this post I will be covering how to extend an [Inception Pipeline](https://github.com/MechanicalRock/InceptionPipeline/tree/master) to do something useful. In this instance, it is creating the infrastructure to host a single page application. On the projects I'm currently involved with, this is always the first piece of infrastructure we need (well, after first inceptioning up the pipeline).
+
+## What Technologies Are We Going to Use
```
Comment (jgunnink) [inline _posts/2018-04-01-inception-pipelines-pt2.markdown]: inconsistent capitalisation

File: _posts/2018-04-01-inception-pipelines-pt2.markdown [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+          InputArtifacts:
+            - Name: !Join ['', [!Ref RepositoryName, 'Source']]
+          RunOrder: '10'
+    ```
+
+4. Copy the `aws_infrastructure.yml` and `aws_infrastructure.json` into the same folder as `aws_seed.yml`
+5. Replace the parameter values in `aws_infrastructure.json` with appropriate values.
+6. Commit the changes and push to CodeCommit.
+
+If you sit and watch the pipeline execution, you'll notice its starts executing, realises there is a structural change and then restarts execution from the top. In my not-so-humble opinion, this is one of the best features of an Inception Pipeline; it just manages itself effortlessly.
```
Comment (jgunnink) [inline _posts/2018-04-01-inception-pipelines-pt2.markdown]: 👍

File: _posts/2018-04-01-inception-pipelines-pt2.markdown (hunk 1 above)
Comment (mechanicalpete) [reply _posts/2018-04-01-inception-pipelines-pt2.markdown]: Tim made the same comment on the last post. I originally did that, but the markdown linter kept saying to remove them, so I did.

Personally I think they should be in, but maybe leave them out for consistency sake?

File: _posts/2018-04-01-inception-pipelines-pt2.markdown (hunk 7 above)
Comment (mechanicalpete) [reply _posts/2018-04-01-inception-pipelines-pt2.markdown]: Fixed. Title cased all the titles to match original post formatting.

File: _posts/2018-04-01-inception-pipelines-pt2.markdown (hunk 2 above)
Comment (mechanicalpete) [reply _posts/2018-04-01-inception-pipelines-pt2.markdown]: Removed, thanks.

File: _posts/2018-04-01-inception-pipelines-pt2.markdown (hunk 8 above)
Comment (Hamo69) [reply _posts/2018-04-01-inception-pipelines-pt2.markdown]: 👍

**Comment by Hamo69** [comment]
I haven't built it locally and looked at how it appears, so beware - may need some rejigging

File: _posts/2018-04-01-inception-pipelines-pt2.markdown (hunk 8 above)
Comment (temyers) [reply _posts/2018-04-01-inception-pipelines-pt2.markdown]: +1

**Pull request merged**